05_debug/sysfs/task_05_sysfs.c

```diff
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: MIT and GPL
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/fs.h>
+#include <linux/cdev.h>
+#include <linux/version.h>
+#include <linux/init.h>
+#include <linux/slab.h>
+#include <linux/string.h>
+
+MODULE_DESCRIPTION("Basic module demo: sysfs");
+MODULE_AUTHOR("AlexShlikhta");
+MODULE_VERSION("0.1");
+MODULE_LICENSE("Dual MIT/GPL"); // this affects the kernel behavior
+
+#define CLASS_ATTR(_name, _mode, _show, _store) \
+struct class_attribute class_attr_##_name = __ATTR(_name, _mode, _show, _store)
+
+static struct class *mod_class;
+
+static uint32_t total_calls;
+static uint32_t total_characters_processed;
+static uint32_t characters_converted;
+static char *text_buf = "Bufer is empty:(";
+
+#define LOW_UP_SHIFT	('a' - 'A')
+
+enum low_up_t {
+	LOW_TO_UP = 0,
+	UP_TO_LOW
+};
+
+/**
+ * @brief      Change case of one character from UPPER to lower.
+ * If symbol is in the range from 'A' to 'Z' it's case will be changet,
+ * in other way this func return the symbol
+ *
+ * @param[in]  in    Symbol from ASCII table
+ *
+ * @return     Symbol with changed or not case.
+ */
+char up2low_char(char in)
+{
+	if (in < 'A' || 'Z' < in)
+		return in;
+
+	return in + LOW_UP_SHIFT;
+}
+
+/**
+ * @brief      Change case of one character from lower to UPPER.
+ * If symbol is in the range from 'a' to 'z' it's case will be changet,
+ * in other way this func return the symbol
+ *
+ * @param[in]  in    Symbol from ASCII table
+ *
+ * @return     Symbol with changed or not case.
+ */
+char low2up_char(char in)
+{
+	if (in < 'a' || 'z' < in)
+		return in;
+
+	return in - LOW_UP_SHIFT;
+}
+
+/**
+ * @brief      Change case of string from UPPER case to lower, or backward.
+ * The result depends on 'shift_dir' flag which is enum and have only 2 states.
+ *
+ * @param[in]  in         Input string with different cases
+ * @param[in]  shift_dir  The shift dir
+ *
+ * @return     Whole string in the same case.
+ */
+char *change_case(const char *in, enum low_up_t shift_dir)
+{
+	uint16_t L = strlen(in);
+	uint16_t i;
+	char *res = kzalloc(sizeof(*res) * L, GFP_KERNEL);
```

Comment:
What is the reason for kzalloc? I suppose kalloc is enough?

Reply:
I just used to use kzalloc in any case.

Reply:
What is the difference between kalloc and kzalloc?